### ISSUE TITLE
Tell me to set SDK to >=2.13, "This requires the 'nonfunction-type-aliases' language feature to be enabled"

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#46329]
```
error: This requires the 'nonfunction-type-aliases' language feature to be enabled. (experiment_not_enabled at [json_serializable] test/type_defs/typedef_usage.dart:9)
```

Once a feature is stable, we should point folks to the version of the SDK they need to set in their pubspec.

The language feature verbiage is not helpful.

**Comment by bwilkerson** [comment]
That message is coming from the parser.

**Comment by johnniwinther** [comment]
The tip associated with this error actually says
```
Try updating your pubspec.yaml to set the minimum SDK constraint to 2.13 or higher, and running 'pub get'.
```
but it is not emitted by the analyzer.

@bwilkerson Should we make the analyzer emit the tip as well?

**Comment by bwilkerson** [comment]
> Should we make the analyzer emit the tip as well?

The analyzer already emits this tip.

@kevmoo Where are you seeing this message?

The `dart analyze` command currently produces the following:
```
  error • test.dart:1:11 • This requires the 'nonfunction-type-aliases' language feature to be enabled. Try updating your
          pubspec.yaml to set the minimum SDK constraint to 2.14.0 or higher, and running 'pub get'. •
          experiment_not_enabled
```

while the `dartanalyzer` command produces
```
  error • This requires the 'nonfunction-type-aliases' language feature to be enabled. • test.dart:1:11 • experiment_not_enabled
```

neither of which match the output you provided.

**Comment by kevmoo** [comment]
@bwilkerson – I was seeing it in IntelliJ

**Comment by bwilkerson** [comment]
That's an IntelliJ presentation issue. The correction message was sent to the client and the client displays it in some places (try hovering over the diagnostic in the Dart Analysis view or in the editor), but it appears that it wasn't being displayed in a place that you discovered.

The question is, where would you have expected to have seen it?

**Comment by kevmoo** [comment]
@bwilkerson – I would hope the line that shows up in the IDE to point to the language version

CC @devoncarew and @jwren 

![image](https://user-images.githubusercontent.com/17034/122307995-4d3fc680-cec0-11eb-8954-ca9e58e756a9.png)

**Comment by bwilkerson** [comment]
Possibilities that I can think of include
- adding a new column in the Dart Analysis view's table,
- appending the correction to the message under "Description",
- making the rows multiple lines long and adding the correction on a second line, or
- making the diagnostics expandable (tree table) and putting the correction in as a child of the diagnostic (possibly with the context messages as children as well).

**Comment by kevmoo** [comment]
Could we change the content of the primary message? It's not helpful at all...

**Comment by bwilkerson** [comment]
Probably. That's coming from the `Parser`, so that would be front-end work. Do you have a suggestion?

**Comment by johnniwinther** [comment]
@kevmoo Do you have a suggestion for what wording would have helped in your situation?

**Comment by bwilkerson** [comment]
FWIW, the non-parser messages produced by the analyzer have the form:

> Extension methods weren't supported until version 2.6.0, but this code is required to be able to run on earlier versions.

Rather than using the experiment name it focuses on the version number (for features that have been released, of course).

**Comment by kevmoo** [comment]
@johnniwinther  "This requires Dart X.Y. Please see dart.dev/x/y/z for more information."

**Comment by mit-mit** [comment]
I like the message Kevin suggests, but think we should link to https://dart.dev/guides/language/evolution#language-versioning instead (we can create a shorter go-link if desired).

**Comment by johnniwinther** [comment]
Most messages related to (experimental) features are not generated in the parser but in the ast builders in the analyzer and cfe, respectively, where the context of the library language version is available. The problem with the parser generated messages is that they are somewhat hardcoded and therefore easily go stale. I'm working on a CL to mitigate this, but I'm not sure that I can currently produce the messages mentioned in this issue (anymore?).

@kevmoo Can you create repros for the problems you've seen?

**Comment by kevmoo** [comment]
Absolutely, @johnniwinther – just use `library;` with latest dev in a package that has `sdk: >=2.17.0 <3.0.0` or similar

**Comment by kevmoo** [comment]
```
error: This requires the 'unnamed-libraries' language feature to be enabled. (experiment_not_enabled at [http2] lib/src/async_utils/async_utils.dart:7)
```

with Dart SDK version: 2.19.0-361.0.dev (dev) (Tue Nov 1 15:30:40 2022 -0700) on "macos_arm64" in intellij

**Comment by johnniwinther** [comment]
@kevmoo Do you know how you got the error related to `nonfunction-type-aliases` in the original issue? The error related to `unnamed-libraries` is provided by the ast builder and not the parser. (The message needs to be updated, though).

**Comment by kevmoo** [comment]
...that might have been fixed. 🤷  I just piled on here because I hit the same flavor of issue w/ the latest SDK (and the "naked" library feature)

**Comment by srawlins** [comment]
Yeah I think it would be super beneficial to just slip the version number into the primary message. Instead of:

> This requires the 'nonfunction-type-aliases' language feature to be enabled.

Something like

> This requires a minimum Dart Language version of '2.13.0', which introduces the 'nonfunction-type-aliases' language feature.

And we should have a quick fix as well (do we?) so that users don't have to go read documentation for something that's really not that interesting; knowledge about Dart SDK versions vs Dart language versions hopefully doesn't need to be top-of-mind for the average developer.
